core/src/main/java/io/grpc/internal/ManagedChannelImpl2.java

```diff
@@ -0,0 +1,873 @@
+    if (terminated) {
+      return;
+    }
+    if (shutdown.get() && subchannels.isEmpty() && oobChannels.isEmpty()) {
+      log.log(Level.FINE, "[{0}] Terminated", getLogId());
+      terminated = true;
+      terminatedLatch.countDown();
+      executorPool.returnObject(executor);
+      // Needed for delivering rejections to new calls after OobChannel is terminated.
+      executor = MoreExecutors.directExecutor();
+      scheduledExecutor = timerServicePool.returnObject(scheduledExecutor);
+      // Release the transport factory so that it can deallocate any resources.
+      transportFactory.close();
+    }
+  }
+
+  private class LbHelperImpl extends LoadBalancer2.Helper {
+    LoadBalancer2 lb;
+    final NameResolver nr;
+
+    LbHelperImpl(NameResolver nr) {
+      this.nr = checkNotNull(nr, "NameResolver");
+    }
+
+    @Override
+    public SubchannelImpl createSubchannel(EquivalentAddressGroup addressGroup, Attributes attrs) {
+      checkNotNull(addressGroup, "addressGroup");
+      checkNotNull(attrs, "attrs");
+      ScheduledExecutorService scheduledExecutorCopy = scheduledExecutor;
+      checkState(scheduledExecutorCopy != null,
+          "scheduledExecutor is already cleared. Looks like you are calling this method after "
+          + "you've already shut down");
+      final SubchannelImplImpl subchannel = new SubchannelImplImpl(attrs);
+      final InternalSubchannel internalSubchannel = new InternalSubchannel(
+            addressGroup, authority(), userAgent, backoffPolicyProvider, transportFactory,
+            scheduledExecutorCopy, stopwatchSupplier, channelExecutor,
+            new InternalSubchannel.Callback() {
+              // All callbacks are run in channelExecutor
+              @Override
+              void onTerminated(InternalSubchannel is) {
+                subchannels.remove(is);
+                maybeTerminateChannel();
+              }
+
+              @Override
+              void onStateChange(InternalSubchannel is, ConnectivityStateInfo newState) {
+                if ((newState.getState() == TRANSIENT_FAILURE || newState.getState() == IDLE)) {
+                  nr.refresh();
+                }
+                lb.handleSubchannelState(subchannel, newState);
+              }
+
+              @Override
+              public void onInUse(InternalSubchannel is) {
+                inUseStateAggregator.updateObjectInUse(is, true);
+              }
+
+              @Override
+              public void onNotInUse(InternalSubchannel is) {
+                inUseStateAggregator.updateObjectInUse(is, false);
+              }
+            });
+      subchannel.subchannel = internalSubchannel;
+      log.log(Level.FINE, "[{0}] {1} created for {2}",
+          new Object[] {getLogId(), internalSubchannel.getLogId(), addressGroup});
+      runSerialized(new Runnable() {
+          @Override
+          public void run() {
+            if (terminating) {
+              internalSubchannel.shutdown();
```

Comment:
This `runSerialized()` block looks subtle. The `runSerialized()` does not guarantee that the `Runnable` has executed before it returns. So we could return a subchannel here that is non-yet-shutdown even when terminating. But that subchannel won't (can't?) be used because the delayed transport is already terminated and so has no RPCs and is not accepting any more RPCs (or passing-through any more RPCs).

It seems a comment would go a long way, if nothing else to prevent future bugs.